source/grid/grid_tools.cc

```diff
@@ -4451,6 +4452,299 @@ namespace GridTools
+                  }
+              }
+            else
+              {
+                Assert(false, ExcNotImplemented());
+              }
+          }
+      }
+
+    // if no cells need to be added, then no regularization is necessary. Restore things
+    // as they were before this function was called.
+    if (cells_to_add.size() == 0)
+      {
+        for (; refinement_cycles>0; refinement_cycles--)
+          {
+            for (auto cell: tria.active_cell_iterators())
+              cell->set_coarsen_flag();
+            tria.execute_coarsening_and_refinement();
+          }
+        return;
+      }
+
+    // add the cells that were not marked as skipped
+    for (auto cell : tria.active_cell_iterators())
+      {
+        if (cells_to_remove[cell->active_cell_index()] == false)
+          {
+            CellData<dim> c;
+            for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+              c.vertices[v] = cell->vertex_index(v);
+            c.manifold_id = cell->manifold_id();
+            c.material_id = cell->material_id();
+            cells_to_add.push_back(c);
+          }
+      }
+
+    // Face counter for both dim == 2 and dim == 3
+    typename Triangulation<dim,spacedim>::active_face_iterator
+    face = tria.begin_active_face(),
+    endf = tria.end_face();
+    for (; face != endf; ++face)
+      if ( (face->at_boundary() || face->manifold_id() != numbers::invalid_manifold_id)
+           && faces_to_remove[face->index()] == false)
+        {
+          for (unsigned int l=0; l<GeometryInfo<dim>::lines_per_face; ++l)
+            {
+              CellData<1> line;
+              if (dim == 2)
+                {
+                  for (unsigned int v=0; v<GeometryInfo<1>::vertices_per_cell; ++v)
+                    line.vertices[v] = face->vertex_index(v);
+                  line.boundary_id = face->boundary_id();
+                  line.manifold_id = face->manifold_id();
+                }
+              else
+                {
+                  for (unsigned int v=0; v<GeometryInfo<1>::vertices_per_cell; ++v)
+                    line.vertices[v] = face->line(l)->vertex_index(v);
+                  line.boundary_id = face->line(l)->boundary_id();
+                  line.manifold_id = face->line(l)->manifold_id();
+                }
+              subcelldata_to_add.boundary_lines.push_back(line);
+            }
+          if (dim == 3)
+            {
+              CellData<2> quad;
+              for (unsigned int v=0; v<GeometryInfo<2>::vertices_per_cell; ++v)
+                quad.vertices[v] = face->vertex_index(v);
+              quad.boundary_id = face->boundary_id();
+              quad.manifold_id = face->manifold_id();
+              subcelldata_to_add.boundary_quads.push_back(quad);
+            }
+        }
+    GridTools::delete_unused_vertices(vertices, cells_to_add, subcelldata_to_add);
+    GridReordering<dim,spacedim>::reorder_cells(cells_to_add, true);
```

Comment:
@luca-heltai Do we still need this line? Weren't we only using it in the hope of avoiding negative cell measures?